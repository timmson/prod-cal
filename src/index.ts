import cal from "../calendars/ru"

const getSequencedArray = (length: number, startFrom = 0) =>
	Array.from(Array(length).keys()).map((i) => i + startFrom)

class Calendar {

	public static DAY_WORK = "work"
	public static DAY_HOLIDAY = "holiday"
	public static DAY_WORK_REDUCED = "work_reduced"

	constructor(locale: string) {
		if (locale && locale !== "ru") {
			throw new Error("Locales differ from RU is not supported yet")
		}

	}

	getCalendar(year: number | Date, month?: number, day?: number) {
		if (!year) {
			throw new Error("Year must be not null")
		}

		if (year instanceof Date) {
			day = year.getDate()
			month = year.getMonth() + 1
			year = year.getFullYear()
		}

		if (month) {
			if (day) {
				let dayType = Calendar.DAY_WORK
				if (cal[year.toString()] !== undefined && cal[year.toString()][month - 1] !== undefined) {
					const monthArray = cal[year.toString()][month - 1]
					dayType = monthArray.includes(day.toString()) ? Calendar.DAY_HOLIDAY : dayType
					dayType = monthArray.includes(day + "*") ? Calendar.DAY_WORK_REDUCED : dayType
				}
				return dayType
			}

			return getSequencedArray(new Date(year, month, 0).getDate(), 1).map(day => this.getCalendar(year, month, day))
		}

		return getSequencedArray(12, 1).map(month => this.getCalendar(year, month))
	}
}

export default Calendar