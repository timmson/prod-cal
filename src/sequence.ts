export const getSequencedArray = (start :number, length: number): Array<number> =>
	Array.from(Array(length).keys()).map((i) => i + start)
