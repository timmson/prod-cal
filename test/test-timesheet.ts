import {TimesheetBuilder} from "../src"

describe("Timesheet should ", () => {

	const timesheetBuilder = new TimesheetBuilder("ru")

	test("TimesheetBuilder gets unknown locale and test should throw error", () => {
		expect(() => new TimesheetBuilder("ru1")).toThrow()
	})

	test("return valid timesheet for 2023-11-1 - 2023-11-30", () => {
		const actual = timesheetBuilder.build(new Date("2023-11-1"), new Date("2023-11-30"), [])

		expect(actual).toMatchSnapshot()
		expect(actual.summary).toEqual({days: 21, hours: 167})
	})


	test("return valid timesheet for 2023-9-18 - 2023-12-15", () => {
		const actual = timesheetBuilder.build(new Date("2023-9-18"), new Date("2023-12-15"), [])

		expect(actual).toMatchSnapshot()
		expect(actual.months[0].summary).toEqual({days: 10, hours: 80})
		expect(actual.months[1].summary).toEqual({days: 22, hours: 176})
		expect(actual.months[2].summary).toEqual({days: 21, hours: 167})
		expect(actual.months[3].summary).toEqual({days: 11, hours: 88})
		expect(actual.summary).toEqual({days: 64, hours: 511})
	})

	test("return valid timesheet for 2023-11-21 - 2024-1-22", () => {
		const actual = timesheetBuilder.build(new Date("2023-11-21"), new Date("2024-1-22"), [])

		expect(actual).toMatchSnapshot()
		expect(actual.months[0].summary).toEqual({days: 8, hours: 64})
		expect(actual.months[1].summary).toEqual({days: 21, hours: 168})
		expect(actual.months[2].summary).toEqual({days: 10, hours: 80})
		expect(actual.summary).toEqual({days: 39, hours: 312})
	})

	test("return valid timesheet for November 2023 with out of office for 20-24", () => {
		const actual = timesheetBuilder.build(new Date("2023-11-1"), new Date("2023-11-30"), [
			{date: new Date(2023, 11-1, 20), hours: 0},
			{date: new Date(2023, 11-1, 21), hours: 0},
			{date: new Date(2023, 11-1, 22), hours: 0},
			{date: new Date(2023, 11-1, 23), hours: 0},
			{date: new Date(2023, 11-1, 24), hours: 0}
		])

		expect(actual).toMatchSnapshot()
		expect(actual.summary).toEqual({days: 16, hours: 127})
	})

})
