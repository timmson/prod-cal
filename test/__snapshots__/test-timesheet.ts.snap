// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timesheet should  return valid timesheet for 2023-9-18 - 2023-12-15 1`] = `
{
  "months": [
    {
      "days": [
        {
          "date": 18,
          "hours": 8,
        },
        {
          "date": 19,
          "hours": 8,
        },
        {
          "date": 20,
          "hours": 8,
        },
        {
          "date": 21,
          "hours": 8,
        },
        {
          "date": 22,
          "hours": 8,
        },
        {
          "date": 23,
          "hours": 0,
        },
        {
          "date": 24,
          "hours": 0,
        },
        {
          "date": 25,
          "hours": 8,
        },
        {
          "date": 26,
          "hours": 8,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 0,
        },
      ],
      "month": 9,
      "summary": {
        "days": 10,
        "hours": 80,
      },
      "year": 2023,
    },
    {
      "days": [
        {
          "date": 1,
          "hours": 0,
        },
        {
          "date": 2,
          "hours": 8,
        },
        {
          "date": 3,
          "hours": 8,
        },
        {
          "date": 4,
          "hours": 8,
        },
        {
          "date": 5,
          "hours": 8,
        },
        {
          "date": 6,
          "hours": 8,
        },
        {
          "date": 7,
          "hours": 0,
        },
        {
          "date": 8,
          "hours": 0,
        },
        {
          "date": 9,
          "hours": 8,
        },
        {
          "date": 10,
          "hours": 8,
        },
        {
          "date": 11,
          "hours": 8,
        },
        {
          "date": 12,
          "hours": 8,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 0,
        },
        {
          "date": 15,
          "hours": 0,
        },
        {
          "date": 16,
          "hours": 8,
        },
        {
          "date": 17,
          "hours": 8,
        },
        {
          "date": 18,
          "hours": 8,
        },
        {
          "date": 19,
          "hours": 8,
        },
        {
          "date": 20,
          "hours": 8,
        },
        {
          "date": 21,
          "hours": 0,
        },
        {
          "date": 22,
          "hours": 0,
        },
        {
          "date": 23,
          "hours": 8,
        },
        {
          "date": 24,
          "hours": 8,
        },
        {
          "date": 25,
          "hours": 8,
        },
        {
          "date": 26,
          "hours": 8,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 0,
        },
        {
          "date": 29,
          "hours": 0,
        },
        {
          "date": 30,
          "hours": 8,
        },
        {
          "date": 31,
          "hours": 8,
        },
      ],
      "month": 10,
      "summary": {
        "days": 22,
        "hours": 176,
      },
      "year": 2023,
    },
    {
      "days": [
        {
          "date": 1,
          "hours": 8,
        },
        {
          "date": 2,
          "hours": 8,
        },
        {
          "date": 3,
          "hours": 7,
        },
        {
          "date": 4,
          "hours": 0,
        },
        {
          "date": 5,
          "hours": 0,
        },
        {
          "date": 6,
          "hours": 0,
        },
        {
          "date": 7,
          "hours": 8,
        },
        {
          "date": 8,
          "hours": 8,
        },
        {
          "date": 9,
          "hours": 8,
        },
        {
          "date": 10,
          "hours": 8,
        },
        {
          "date": 11,
          "hours": 0,
        },
        {
          "date": 12,
          "hours": 0,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 8,
        },
        {
          "date": 15,
          "hours": 8,
        },
        {
          "date": 16,
          "hours": 8,
        },
        {
          "date": 17,
          "hours": 8,
        },
        {
          "date": 18,
          "hours": 0,
        },
        {
          "date": 19,
          "hours": 0,
        },
        {
          "date": 20,
          "hours": 8,
        },
        {
          "date": 21,
          "hours": 8,
        },
        {
          "date": 22,
          "hours": 8,
        },
        {
          "date": 23,
          "hours": 8,
        },
        {
          "date": 24,
          "hours": 8,
        },
        {
          "date": 25,
          "hours": 0,
        },
        {
          "date": 26,
          "hours": 0,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 8,
        },
      ],
      "month": 11,
      "summary": {
        "days": 21,
        "hours": 167,
      },
      "year": 2023,
    },
    {
      "days": [
        {
          "date": 1,
          "hours": 8,
        },
        {
          "date": 2,
          "hours": 0,
        },
        {
          "date": 3,
          "hours": 0,
        },
        {
          "date": 4,
          "hours": 8,
        },
        {
          "date": 5,
          "hours": 8,
        },
        {
          "date": 6,
          "hours": 8,
        },
        {
          "date": 7,
          "hours": 8,
        },
        {
          "date": 8,
          "hours": 8,
        },
        {
          "date": 9,
          "hours": 0,
        },
        {
          "date": 10,
          "hours": 0,
        },
        {
          "date": 11,
          "hours": 8,
        },
        {
          "date": 12,
          "hours": 8,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 8,
        },
        {
          "date": 15,
          "hours": 8,
        },
      ],
      "month": 12,
      "summary": {
        "days": 11,
        "hours": 88,
      },
      "year": 2023,
    },
  ],
  "summary": {
    "days": 64,
    "hours": 511,
  },
}
`;

exports[`Timesheet should  return valid timesheet for 2023-11-1 - 2023-11-30 1`] = `
{
  "months": [
    {
      "days": [
        {
          "date": 1,
          "hours": 8,
        },
        {
          "date": 2,
          "hours": 8,
        },
        {
          "date": 3,
          "hours": 7,
        },
        {
          "date": 4,
          "hours": 0,
        },
        {
          "date": 5,
          "hours": 0,
        },
        {
          "date": 6,
          "hours": 0,
        },
        {
          "date": 7,
          "hours": 8,
        },
        {
          "date": 8,
          "hours": 8,
        },
        {
          "date": 9,
          "hours": 8,
        },
        {
          "date": 10,
          "hours": 8,
        },
        {
          "date": 11,
          "hours": 0,
        },
        {
          "date": 12,
          "hours": 0,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 8,
        },
        {
          "date": 15,
          "hours": 8,
        },
        {
          "date": 16,
          "hours": 8,
        },
        {
          "date": 17,
          "hours": 8,
        },
        {
          "date": 18,
          "hours": 0,
        },
        {
          "date": 19,
          "hours": 0,
        },
        {
          "date": 20,
          "hours": 8,
        },
        {
          "date": 21,
          "hours": 8,
        },
        {
          "date": 22,
          "hours": 8,
        },
        {
          "date": 23,
          "hours": 8,
        },
        {
          "date": 24,
          "hours": 8,
        },
        {
          "date": 25,
          "hours": 0,
        },
        {
          "date": 26,
          "hours": 0,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 8,
        },
      ],
      "month": 11,
      "summary": {
        "days": 21,
        "hours": 167,
      },
      "year": 2023,
    },
  ],
  "summary": {
    "days": 21,
    "hours": 167,
  },
}
`;

exports[`Timesheet should  return valid timesheet for 2023-11-21 - 2024-1-22 1`] = `
{
  "months": [
    {
      "days": [
        {
          "date": 21,
          "hours": 8,
        },
        {
          "date": 22,
          "hours": 8,
        },
        {
          "date": 23,
          "hours": 8,
        },
        {
          "date": 24,
          "hours": 8,
        },
        {
          "date": 25,
          "hours": 0,
        },
        {
          "date": 26,
          "hours": 0,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 8,
        },
      ],
      "month": 11,
      "summary": {
        "days": 8,
        "hours": 64,
      },
      "year": 2023,
    },
    {
      "days": [
        {
          "date": 1,
          "hours": 8,
        },
        {
          "date": 2,
          "hours": 0,
        },
        {
          "date": 3,
          "hours": 0,
        },
        {
          "date": 4,
          "hours": 8,
        },
        {
          "date": 5,
          "hours": 8,
        },
        {
          "date": 6,
          "hours": 8,
        },
        {
          "date": 7,
          "hours": 8,
        },
        {
          "date": 8,
          "hours": 8,
        },
        {
          "date": 9,
          "hours": 0,
        },
        {
          "date": 10,
          "hours": 0,
        },
        {
          "date": 11,
          "hours": 8,
        },
        {
          "date": 12,
          "hours": 8,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 8,
        },
        {
          "date": 15,
          "hours": 8,
        },
        {
          "date": 16,
          "hours": 0,
        },
        {
          "date": 17,
          "hours": 0,
        },
        {
          "date": 18,
          "hours": 8,
        },
        {
          "date": 19,
          "hours": 8,
        },
        {
          "date": 20,
          "hours": 8,
        },
        {
          "date": 21,
          "hours": 8,
        },
        {
          "date": 22,
          "hours": 8,
        },
        {
          "date": 23,
          "hours": 0,
        },
        {
          "date": 24,
          "hours": 0,
        },
        {
          "date": 25,
          "hours": 8,
        },
        {
          "date": 26,
          "hours": 8,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 0,
        },
        {
          "date": 31,
          "hours": 0,
        },
      ],
      "month": 12,
      "summary": {
        "days": 21,
        "hours": 168,
      },
      "year": 2023,
    },
    {
      "days": [
        {
          "date": 1,
          "hours": 0,
        },
        {
          "date": 2,
          "hours": 0,
        },
        {
          "date": 3,
          "hours": 0,
        },
        {
          "date": 4,
          "hours": 0,
        },
        {
          "date": 5,
          "hours": 0,
        },
        {
          "date": 6,
          "hours": 0,
        },
        {
          "date": 7,
          "hours": 0,
        },
        {
          "date": 8,
          "hours": 0,
        },
        {
          "date": 9,
          "hours": 8,
        },
        {
          "date": 10,
          "hours": 8,
        },
        {
          "date": 11,
          "hours": 8,
        },
        {
          "date": 12,
          "hours": 8,
        },
        {
          "date": 13,
          "hours": 0,
        },
        {
          "date": 14,
          "hours": 0,
        },
        {
          "date": 15,
          "hours": 8,
        },
        {
          "date": 16,
          "hours": 8,
        },
        {
          "date": 17,
          "hours": 8,
        },
        {
          "date": 18,
          "hours": 8,
        },
        {
          "date": 19,
          "hours": 8,
        },
        {
          "date": 20,
          "hours": 0,
        },
        {
          "date": 21,
          "hours": 0,
        },
        {
          "date": 22,
          "hours": 8,
        },
      ],
      "month": 1,
      "summary": {
        "days": 10,
        "hours": 80,
      },
      "year": 2024,
    },
  ],
  "summary": {
    "days": 39,
    "hours": 312,
  },
}
`;

exports[`Timesheet should  return valid timesheet for November 2023 with out of office for 20-24 1`] = `
{
  "months": [
    {
      "days": [
        {
          "date": 1,
          "hours": 8,
        },
        {
          "date": 2,
          "hours": 8,
        },
        {
          "date": 3,
          "hours": 7,
        },
        {
          "date": 4,
          "hours": 0,
        },
        {
          "date": 5,
          "hours": 0,
        },
        {
          "date": 6,
          "hours": 0,
        },
        {
          "date": 7,
          "hours": 8,
        },
        {
          "date": 8,
          "hours": 8,
        },
        {
          "date": 9,
          "hours": 8,
        },
        {
          "date": 10,
          "hours": 8,
        },
        {
          "date": 11,
          "hours": 0,
        },
        {
          "date": 12,
          "hours": 0,
        },
        {
          "date": 13,
          "hours": 8,
        },
        {
          "date": 14,
          "hours": 8,
        },
        {
          "date": 15,
          "hours": 8,
        },
        {
          "date": 16,
          "hours": 8,
        },
        {
          "date": 17,
          "hours": 8,
        },
        {
          "date": 18,
          "hours": 0,
        },
        {
          "date": 19,
          "hours": 0,
        },
        {
          "date": 20,
          "hours": 0,
        },
        {
          "date": 21,
          "hours": 0,
        },
        {
          "date": 22,
          "hours": 0,
        },
        {
          "date": 23,
          "hours": 0,
        },
        {
          "date": 24,
          "hours": 0,
        },
        {
          "date": 25,
          "hours": 0,
        },
        {
          "date": 26,
          "hours": 0,
        },
        {
          "date": 27,
          "hours": 8,
        },
        {
          "date": 28,
          "hours": 8,
        },
        {
          "date": 29,
          "hours": 8,
        },
        {
          "date": 30,
          "hours": 8,
        },
      ],
      "month": 11,
      "summary": {
        "days": 16,
        "hours": 127,
      },
      "year": 2023,
    },
  ],
  "summary": {
    "days": 16,
    "hours": 127,
  },
}
`;
